### PR TITLE
Allow requests with cors specified in a list of regexes or urls and setup heroku

### LINKED ISSUE #145: Add CORS config help to README
Add some explanation to what should be placed in the `ACCESS_CONTROL_ALLOW_ORIGIN` variable

### PR DESCRIPTION
**Pull request by DoStini**
We are trying to give a more interesting experience for frontend previews.
In heroku deploys (which we will be using also for backend previews), we will allow any request from one of our frontend previews

closes #145

### CODE REVIEW

**Comment by DoStini** [comment]
@imnotteixeira @miguelpduarte Do you see any problems with this approach? Do you think there is a different way of doing dynamic cors?

Cors with "*" didn't work due to the use of `credentials="include"` in frontend requests

I've also checked and netlify doesn't allow the creation of  links with this regex

image

File: src/config/env.js [hunk 1]
```diff
@@ -30,6 +30,10 @@ export default Object.freeze({
     admin_password: process.env.ADMIN_PASSWORD,
     access_control_allow_origin: process.env.ACCESS_CONTROL_ALLOW_ORIGIN || "*",
 
+    // Deploy previews
+    is_heroku: process.env.IS_HEROKU,
```
Comment (imnotteixeira) [inline src/config/env.js:34]: This variable is not needed, simply have an allowlist of allowed domains to query the api

**Review by imnotteixeira — CHANGES_REQUESTED** [review]

File: src/config/env.js (hunk 1 above)
Comment (imnotteixeira) [reply src/config/env.js:34]: Also, the name would be misleading, this should not be coupled with any deployment service. What if we use vercel, netlify, digital ocean/ something else?

**Comment by imnotteixeira** [comment]
> @imnotteixeira @miguelpduarte Do you see any problems with this approach? Do you think there is a different way of doing dynamic cors?
> 
> Cors with "*" didn't work due to the use of `credentials="include"` in frontend requests
> 
> I've also checked and netlify doesn't allow the creation of links with this regex
> 
> image

As you've noticed, allowing `*` does not work due to `credentials=include`, as well as not being a good practice.

The way to go is what you've done here: have an allowlist of domains, and only return the header if it matches the allowlisted domain

File: src/config/env.js (hunk 1 above)
Comment (DoStini) [reply src/config/env.js:34]: All right then

Comment (DoStini) [reply src/config/env.js]: I believe this is a good implementation to what we want.

We will be using this regex in beta deploys (or heroku): `"https:\/\/deploy-preview-\d+--nijobs\.netlify\.app"`

Comment (imnotteixeira) [reply src/config/env.js]: I don't think a regex is the way to go here. It solves the very specific problem we have now, sure, but what if there's a new provider we may want to use as well as netlify, what about local frontend builds? No one should have to know/understand this regex in order to allowlist something.

I think the ideal solution would be to make this variable a JSON string representing an array of regexes. It allows the solution you have now, but people can still add more domains easily, without having to think about regexes (exact strings are still valid regexes anyway)

In the end, the config object would parse the JSON into a JS array of Regexes, which can be evaluated against the requester's origin

Comment (DoStini) [reply src/config/env.js]: I think this might be the way to go. I left some examples in the readme and in the env file

File: README.md [hunk 2]
```diff
@@ -40,6 +39,14 @@ To start developing, copy `.env` to `.env.local` (by running `cp .env .env.local
...
 Then, you can override the variable's values, according to their explanation in `.env`.
 
+#### CORS Setup
+
+In order to allow requests from multiple origins besides the one defined by `ACCESS_CONTROL_ALLOW_ORIGIN`, an array of URL's or regexes can be defined by `ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST`. It is crucial that all URL's have no trailing `/`. Example:
+
+```
+ACCESS_CONTROL_ALLOW_ORIGIN=https://localhost:3000
+ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST=["https:\\/\\/deploy-preview-\\d+--nijobs\\.netlify\\.app", "https://nijobs.netlify.app"]
+```
```
Comment (Naapperas) [inline README.md:49]: Couldn't this be used to close #145 ?

Comment (DoStini) [reply README.md:49]: Yup

File: .env [hunk 3]
```diff
@@ -32,8 +32,12 @@ TEST_LOG_REQUESTS=false
 # Default admin credentials - ALSO OVERRIDE IN PRODUCTION
 ADMIN_EMAIL=ni@aefeup.pt
 ADMIN_PASSWORD=n1j0bs_ftw.12345
+
 #CORS allowed origin - OVERRIDE IN PRODUCTION
 ACCESS_CONTROL_ALLOW_ORIGIN=
```
Comment (imnotteixeira) [inline .env:37]: Remove this var and just use the new one (remember to edit the config/env.js)

File: .env [hunk 4]
```diff
@@ -32,8 +32,12 @@ TEST_LOG_REQUESTS=false
 # Default admin credentials - ALSO OVERRIDE IN PRODUCTION
 ADMIN_EMAIL=ni@aefeup.pt
 ADMIN_PASSWORD=n1j0bs_ftw.12345
+
 #CORS allowed origin - OVERRIDE IN PRODUCTION
 ACCESS_CONTROL_ALLOW_ORIGIN= 
+# List of regexes or url's specifying allowed origins. Example:
+# ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST=["https:\\/\\/deploy-preview-\\d+--nijobs\\.netlify\\.app", "https://nijobs.netlify.app"]
+ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST=
```
Comment (imnotteixeira) [inline .env:40]:
```suggestion
ACCESS_CONTROL_ALLOW_ORIGINS=
```

File: README.md [hunk 5]
```diff
@@ -40,6 +39,14 @@ To start developing, copy `.env` to `.env.local` (by running `cp .env .env.local
 
 Then, you can override the variable's values, according to their explanation in `.env`.
 
+#### CORS Setup
+
+In order to allow requests from multiple origins besides the one defined by `ACCESS_CONTROL_ALLOW_ORIGIN`, an array of URL's or regexes can be defined by `ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST`. It is crucial that all URL's have no trailing `/`. Example:
+
+```
+ACCESS_CONTROL_ALLOW_ORIGIN=https://localhost:3000
```
Comment (imnotteixeira) [inline README.md:47]: No need for the first one

File: src/loaders/express.js [hunk 6]
```diff
@@ -76,9 +76,23 @@ export default (app) => {
...
-        // Allow connections for all origins
+    app.use((req, res, next) => {
+        // Allow connections for default specified connections
         res.setHeader("Access-Control-Allow-Origin", config.access_control_allow_origin);
+
+        // Allow requests from connections specified by the allow list regexes
+        const origin = req.header("origin")?.toLowerCase();
+        if (origin) {
+            const originAllowed = config.access_control_allow_origin_regex_list.find((allowOrigin) => {
+                const matches = origin.match(RegExp(allowOrigin, "g"));
```
Comment (imnotteixeira) [inline src/loaders/express.js:87]: There's probably a built-in way to just check if a particular string matches a regex, no need to get the matches, maybe .test or .matches or something, can't remember.

**Review by imnotteixeira — CHANGES_REQUESTED** [review]

File: .env (hunk 3 above)
Comment (DoStini) [reply .env:37]: @BrunoRosendo We'll have to update .env in our server

**Comment by BrunoRosendo** [comment]
@DoStini Can this be implemented with another service? It'd be very useful to have previews for the backend

File: .env (hunk 3 above)
Comment (imnotteixeira) [inline .env]: Why do we need both settings? We can just use `ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST` now.

File: documentation/docs/intro/getting-started.md [hunk 7]
```diff
@@ -34,6 +34,16 @@ To start developing, copy `.env` to `.env.local` (by running `cp .env .env.local
...
 
+Whenever the file is updated, the server needs to be restarted for changes to have effect.
+
+
+### Cors Setup
+
+In order to allow requests from multiple clients, an array of URL's or regexes can be defined by `ACCESS_CONTROL_ALLOW_ORIGINS`. This is an important step regarding connecting the frontend and the backend of the project. It is crucial that all URL's have no trailing `/`. Example:
+
+```
+ACCESS_CONTROL_ALLOW_ORIGINS=["https:\/\/localhost:3000", "https:\\/\\/deploy-preview-\\d+--nijobs\\.netlify\\.app", "https://nijobs.netlify.app"]
```
Comment (imnotteixeira) [inline documentation/docs/intro/getting-started.md:45]: Maybe you'll want to rename to `ACCESS_CONTROL_ALLOW_ORIGIN_REGEX_LIST`

File: src/loaders/express.js [hunk 8]
```diff
@@ -49,7 +49,7 @@ export default (app) => {
 
     // Set the API call rate limit only on production
     if (process.env.NODE_ENV === "production") {
-        const api_rate_limiter = new RateLimit({
+        const api_rate_limiter = RateLimit({
```
Comment (imnotteixeira) [inline src/loaders/express.js:52]: Why this change?

File: src/loaders/express.js [hunk 9]
```diff
@@ -76,9 +76,23 @@ export default (app) => {
     }
 
     // Adding headers (CORS)
-    app.use((_, res, next) => {
-        // Allow connections for all origins
-        res.setHeader("Access-Control-Allow-Origin", config.access_control_allow_origin);
+    app.use((req, res, next) => {
+        // Allow connections for default specified connections
```
Comment (imnotteixeira) [inline src/loaders/express.js]: Comment seems wrong

File: src/loaders/express.js [hunk 10]
```diff
@@ -76,9 +76,23 @@ export default (app) => {
...
+        res.setHeader("Access-Control-Allow-Origin", "null");
+
+        // Allow requests from connections specified by the allow list regexes
+        const origin = req.header("origin")?.toLowerCase();
```
Comment (imnotteixeira) [inline src/loaders/express.js]: Why convert to lower case? Might cause some misunderstandings with the regex provided in the config file

File: src/loaders/express.js [hunk 11]
```diff
@@ -91,7 +105,7 @@ export default (app) => {
     });
 
     // Health check endpoint
-    app.get("/", (_, res) => res.status(HTTPStatus.OK).json({ "online": true }));
+    app.get("/", (_, res) => res.status(HTTPStatus.OK).send("OK"));
```
Comment (imnotteixeira) [inline src/loaders/express.js]: Why?

**Review by imnotteixeira — CHANGES_REQUESTED** [review]

File: .env (hunk 3 above)
Comment (DoStini) [reply .env]: We don't, forgot to remove it from there

File: src/loaders/express.js (hunk 8 above)
Comment (DoStini) [reply src/loaders/express.js:52]: It isn't working correctly and crashed the app in production mode. There was probably an update to this library

File: documentation/docs/intro/getting-started.md (hunk 7 above)
Comment (DoStini) [reply documentation/docs/intro/getting-started.md:45]: You previously requested to change to `origins`

Comment (imnotteixeira) [reply documentation/docs/intro/getting-started.md:45]: Sure, I meant that it needs to match whatever is on .env

File: src/loaders/express.js (hunk 11 above)
Comment (DoStini) [reply src/loaders/express.js]: I think Render required the health check endpoint to return this, I will try again with the previous version

File: documentation/docs/intro/getting-started.md (hunk 7 above)
Comment (DoStini) [reply documentation/docs/intro/getting-started.md:45]: It is the same now

**Review by imnotteixeira — APPROVED** [review]
